From 284b60ee7762159481432f76efb54f8991870d8b Mon Sep 17 00:00:00 2001
From: zrj <rimvydas.jasinskas@gmail.com>
Date: Mon, 26 Aug 2019 21:58:33 +0300
Subject: [PATCH 56/66] world: Reduce pthread type namespace pollution.

 This fully removes pthread namespace in kernel (except vkernel).
 The vkernel bits first require <signal.h> exposure adjustments.
 * Switch to lighter includes in several headers where possible.
 * Adjust visibility in <signal.h>.
 * Exclude type of sigev_notify_attributes for kernel, used only by the
   lib/librt/aio.c.  We could use void type unconditionally like NetBSD,
 * Move pthread type inclusion to the end of <sys/types.h>, better asm.
 * Do not provide pthread_attr_t in <sys/aio.h> for kernel.
 * Remove <sys/signal.h> from <pthead.h>, seems to be legacy remnant for
   no longer required pthread_kill() and pthread_sigmask().
---
 include/pthread.h    |  3 +--
 include/pthread_np.h |  2 +-
 include/signal.h     |  5 +++--
 sys/sys/aio.h        |  2 ++
 sys/sys/signal.h     | 13 ++++++++++---
 sys/sys/types.h      |  9 ++++++---
 6 files changed, 23 insertions(+), 11 deletions(-)

diff --git a/include/pthread.h b/include/pthread.h
index a2192dfb5c..9252e94101 100644
--- a/include/pthread.h
+++ b/include/pthread.h
@@ -39,9 +39,8 @@
  * Header files.
  */
 #include <sys/cdefs.h>
-#include <sys/types.h>
+#include <sys/_pthreadtypes.h>
 #include <sys/time.h>
-#include <sys/signal.h>
 #include <machine/limits.h>
 #include <sys/sched.h>
 
diff --git a/include/pthread_np.h b/include/pthread_np.h
index af52f0a7f9..28a685785e 100644
--- a/include/pthread_np.h
+++ b/include/pthread_np.h
@@ -32,7 +32,7 @@
 #define _PTHREAD_NP_H_
 
 #include <sys/cdefs.h>
-#include <sys/types.h>
+#include <sys/_pthreadtypes.h>
 #include <sys/cpumask.h>
 #include <time.h>
 
diff --git a/include/signal.h b/include/signal.h
index a6930e8bc3..e9448e877b 100644
--- a/include/signal.h
+++ b/include/signal.h
@@ -37,7 +37,6 @@
 #include <sys/cdefs.h>
 #include <sys/signal.h>
 #include <sys/time.h>
-#include <sys/types.h>
 
 #if __BSD_VISIBLE
 extern const char * const sys_signame[NSIG];
@@ -48,11 +47,13 @@ extern const int sys_nsig;
 __BEGIN_DECLS
 int	raise(int);
 
-#if __POSIX_VISIBLE
+#if __POSIX_VISIBLE || __XSI_VISIBLE
 int	kill(pid_t, int);
+#if __POSIX_VISIBLE >= 199506 || __XSI_VISIBLE >= 500
 int	pthread_kill(pthread_t, int);
 int	pthread_sigmask(int, const sigset_t * __restrict,
 	    sigset_t * __restrict);
+#endif
 int	sigaction(int, const struct sigaction * __restrict,
 	    struct sigaction * __restrict);
 int	sigaddset(sigset_t *, int);
diff --git a/sys/sys/aio.h b/sys/sys/aio.h
index aca1f5a55d..a135cc8212 100644
--- a/sys/sys/aio.h
+++ b/sys/sys/aio.h
@@ -19,7 +19,9 @@
 #ifndef _SYS_AIO_H_
 #define	_SYS_AIO_H_
 
+#ifndef _KERNEL
 #include <sys/_pthreadtypes.h>
+#endif
 #include <sys/_timespec.h>
 #include <sys/signal.h>
 #include <sys/stdint.h>
diff --git a/sys/sys/signal.h b/sys/sys/signal.h
index c467339546..1c5168451b 100644
--- a/sys/sys/signal.h
+++ b/sys/sys/signal.h
@@ -39,7 +39,6 @@
 #define	_SYS_SIGNAL_H_
 
 #include <sys/cdefs.h>
-#include <sys/_pthreadtypes.h>
 #include <sys/_sigset.h>
 #include <sys/stdint.h>		/* for __ types */
 
@@ -237,13 +236,21 @@ union sigval {
 };
 #endif
 
-#if __POSIX_VISIBLE >= 199309
+#if __POSIX_VISIBLE >= 199309 || __XSI_VISIBLE >= 500
+#if !defined(_KERNEL) || defined(_KERNEL_VIRTUAL)
+#include <sys/_pthreadtypes.h>
+#endif
+
 struct sigevent {
 	int	sigev_notify;		/* Notification type */
 	union {
 		int	__sigev_signo;	/* Signal number */
 		int	__sigev_notify_kqueue;
+#ifdef _KERNEL
+		void	*__sigev_notify_attributes;
+#else
 		pthread_attr_t *__sigev_notify_attributes;
+#endif
 	} __sigev_u;
 	union sigval sigev_value;	/* Signal value */
 	void (*sigev_notify_function)(union sigval);
@@ -291,7 +298,7 @@ typedef	struct __sigset	sigset_t;
 /*
  * XXX - there are some nasty dependencies on include file order. Now that
  * sigset_t has been defined we can include the MD header.
- */     
+ */
 #include <machine/signal.h>     /* sig_atomic_t; trap codes; sigcontext */
 
 #if __POSIX_VISIBLE
diff --git a/sys/sys/types.h b/sys/sys/types.h
index 4d8f3f490b..d03a78c40c 100644
--- a/sys/sys/types.h
+++ b/sys/sys/types.h
@@ -51,9 +51,6 @@
 #ifndef _SYS_STDINT_H_
 #include <sys/stdint.h>
 #endif
-#ifndef _SYS__PTHREADTYPES_H_
-#include <sys/_pthreadtypes.h>
-#endif
 
 #if __BSD_VISIBLE
 typedef	unsigned char	u_char;
@@ -242,6 +239,12 @@ typedef	__time_t	time_t;
 typedef	__timer_t	timer_t;
 #endif
 
+#ifndef _KERNEL
+#if __POSIX_VISIBLE >= 199209 || __XSI_VISIBLE >= 500
+#include <sys/_pthreadtypes.h>		/* now POSIX thread types */
+#endif
+#endif /* !_KERNEL */
+
 #if __BSD_VISIBLE
 
 #include <sys/_fd_set.h>
-- 
2.22.0

